node/src/proxy_server/mod.rs

```diff
@@ -278,23 +279,24 @@ impl ProxyServer {
                 })
                 .clone()
         };
-        let host_name = match &return_route_info.server_name {
+        let host_name = match &return_route_info.server_name_opt {
             Some(name) => name.clone(),
-            None => "<unspecified server>".to_string(),
+            None => UNSPECIFIED_SERVER.to_string(),
         };
         let response = &msg.payload;
         match self.keys_and_addrs.a_to_b(&response.stream_key) {
             Some(socket_addr) => {
-                self.subs
-                    .as_ref()
-                    .expect("Neighborhood unbound in ProxyServer")
-                    .update_node_record_metadata
-                    .try_send(NodeRecordMetadataMessage {
-                        public_key: exit_public_key.clone(),
-                        unreachable_host_name_opt: Some(host_name),
-                    })
-                    .expect("Neighborhood is dead");
-
+                if host_name.ne(UNSPECIFIED_SERVER) {
+                    self.subs
+                        .as_ref()
+                        .expect("Neighborhood unbound in ProxyServer")
+                        .update_node_record_metadata
+                        .try_send(NodeRecordMetadataMessage {
+                            public_key: exit_public_key.clone(),
+                            metadata_change: NRMetadataChange::AddUnreachableHost { host_name },
+                        })
+                        .expect("Neighborhood is dead");
+                }
```

Comment:
This is still unpleasantly unclear. I understand what you're trying to solve but I'm nervous from our small knowledge of the wider mechanics. I have a hypothesis that the situation for which we are creating a new condition and an optional string might never happen in reality. The hypothesis is simply based on the logical reasoning that if something doesn't have a host name it then implies need of a substitute, the already workable IP address. With that in mind, we probably shouldn't get a notification about `DNSResolutionFailure` if it has been the pure IP address which traveled to the Exit Node and thus it shouldn't perform a the process of DNS resolution and thus if the process wasn't started it also cannot end with the DNS lookup failure. 

So I sought for the trajectory of the error backwards, from your ProxyServer across Hoper and ProxyClient whose part, the StreamHandlerPool actually, if I hasn't been mistaken, faces decision making in `fn make_stream_with_key(...)` function, where one of the options is to go to `Self::lookup_dns(...)` and that is the place, I believe, where a failure of the process brings about the initial impulse of handling the DNS lookup failure by firing `DnsResolveFailure_0v1` out.

I do think we should inspect carefully this place (very likely mainly `make_stream_with_key`) because I think it can blow off the doubts. If it's true that a hostname-less request can never produce this behavior then it means we treat it pointlessly and we have tests for all this might be confusing for our future work, so I'd prefer to have a serious focus on it now.      

I'm not able to say that it's 100% true because I've should have been sleeping already and my body is protesting, but I'd also like to think about this collectively. Best also with @dnwiebe

Reply:
The `match` statement explained below (it is being used to solve the problem of the current card) existed before I started working on this card.

```rust
        let host_name = match &return_route_info.server_name_opt {
            Some(name) => name.clone(),
            None => UNSPECIFIED_SERVER.to_string(),
        };
```

I was convinced that the author of this code considered a case when this might happen, and the `None` arm might execute. I'm unsure whether it was a necessity or a good to have thing.

When I started working on this card, I moved this `match` statement to few lines above so that it can be used at two places (use case of this card as well as its original use case).

So, basically using the optional string was something that's been used way before this card was implemented. I'm unsure whether it's still desirable or not? This is the question that I consider others can answer better than me.